add lock for ncclCommAbort

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31901 add lock for ncclCommAbort**

ncclCommAbort is not thread safe, so adding a lock for it

Differential Revision: [D19293869](https://our.internmc.facebook.com/intern/diff/D19293869/)

zhaojuanmao committed:
add lock for ncclCommAbort

ncclCommAbort is not thread safe, so adding a lock for it

Differential Revision: [D19293869](https://our.internmc.facebook.com/intern/diff/D19293869/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 0e218af3:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=94071371027fba66f5d55810b06333f5e75c3f6c&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base 94071371 since Jan 03
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  Since your merge base is older than `viable/strict`, run these commands:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

zhaojuanmao committed:
Update on "add lock for ncclCommAbort"


ncclCommAbort is not thread safe, so adding a lock for it

Differential Revision: [D19293869](https://our.internmc.facebook.com/intern/diff/D19293869/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/NCCLUtils.hpp:
```diff
@@ -132,6 +136,7 @@ class NCCLComm {
 
   ncclResult_t checkForNcclError() {
```
We need a lock for `isAborted` as well.

pritamdamania87 commented on torch/lib/c10d/NCCLUtils.hpp:
```diff
@@ -92,13 +93,15 @@ class NCCLComm {
 
   // Move assignable
   NCCLComm& operator=(NCCLComm&& other) {
```
Don't we need a lock in the move constructor above? And also in the destructor?

pietern replied:
Move assignment means we'd need to lock both instances, not just one of them.

I think we need to take a look first where this is moved to begin with, before adding locks everywhere. Also, is move assignment even required for this class? Things would be a lot easier if we could simply remove the move assignment operator.

zhaojuanmao replied:
remove move assignment sounds good to me. How do you think? @pritamdamania87

pritamdamania87 replied:
My feeling is that we should keep the two locks and the move assignment operator since these move constructors and assignment operators were probably added following the [Rule of five](https://en.wikipedia.org/wiki/Rule_of_three_(C%2B%2B_programming)#Rule_of_Five).

zhaojuanmao replied:
I'm removing this move assignment, looks like that simply adding two locks can still cause deadlock unless we lock the two locks as well based on this discussion https://stackoverflow.com/questions/29986208/how-should-i-deal-with-mutexes-in-movable-types-in-c. 

since there is no testing or is not trivial to add testing for this function, and I searched there is no code depending on this, I'm removing it

pritamdamania87 reviewed (changes requested):
Requesting changes since I think we need to lock in a few more places.

zhaojuanmao committed:
Update on "add lock for ncclCommAbort"


ncclCommAbort is not thread safe, so adding a lock for it

Differential Revision: [D19293869](https://our.internmc.facebook.com/intern/diff/D19293869/)

[ghstack-poisoned]

pietern commented on torch/lib/c10d/NCCLUtils.hpp:
```diff
@@ -85,27 +87,31 @@ class NCCLComm {
 
   // Move constructable
   NCCLComm(NCCLComm&& other) {
```
You need to grab the mutex of the other NCCLComm because `this` is being constructed.

pietern commented on torch/lib/c10d/NCCLUtils.hpp:
```diff
@@ -57,6 +58,7 @@ class NCCLComm {
   NCCLComm() : NCCLComm(nullptr) {}
 
   ~NCCLComm() noexcept {
```
Why lock in the destructor?

Are the raw pointer references to this NCCLComm that may cause multiple threads to access it even when it is being destructed? If not, the mutex should not be needed here.

zhaojuanmao replied:
the aborted_ may be accessed by multiple threads?

pritamdamania87 replied:
I suggested having a mutex here since won't have a memory barrier without the mutex. As an example, a thread could call ncclCommAbort setting aborted_ = true, now if a separate thread calls the destructor, the destructor might not see the latest update to aborted_ without a lock (causing it to abort the communicator again).

pietern replied:
If two threads can race on destruction then that's a separate problem.

I would assume that access to the NCCLComm is also somehow protected by a lock (access to a map of some sort), or that it's always wrapped in a shared_ptr. If it's wrapped in a shared_ptr, then by definition, there can only be a single thread calling the destructor.

pritamdamania87 replied:
My feeling is that we shouldn't rely on how the class is used in order to guarantee safety for internal details of this class.

Also, my previous comment wasn't about two threads racing on destruction (I don't think the destructor can be called from two threads). It was more about the fact that the thread calling the destructor wouldn't see the latest value of `aborted_` (which might be updated by separate thread earlier) without a memory barrier.

zhaojuanmao replied:
sounds good to me that we add a lock here to protect aborted_ value is synced up. but agree that we always create a share_ptr<NCCLComm> in this class and will not be accessed by two threads though.

pietern replied:
@pritamdamania87 @zhaojuanmao I see, I think that makes sense. Can you add a comment explaining this rationale to the destructor where the lock is acquired? This discussion is highly relevant and would prevent future removal of the code (as others might reason similarly).

pietern reviewed (changes requested):
Is `ncclCommAbort` not thread safe globally, or only per NCCL communicator?

zhaojuanmao commented:
@pietern thanks for reviewing! IIUC, ncclCommAbort is not thread safe per NCCL communicator. it could be modified by both work thread and watchDog thread.

zhaojuanmao committed:
Update on "add lock for ncclCommAbort"


ncclCommAbort is not thread safe, so adding a lock for it

Differential Revision: [D19293869](https://our.internmc.facebook.com/intern/diff/D19293869/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add lock for ncclCommAbort"


ncclCommAbort is not thread safe, so adding a lock for it

Differential Revision: [D19293869](https://our.internmc.facebook.com/intern/diff/D19293869/)

[ghstack-poisoned]

zhaojuanmao commented:
added comment for destructor.

pritamdamania87 commented on torch/lib/c10d/NCCLUtils.hpp:
```diff
@@ -83,29 +87,27 @@ class NCCLComm {
+
   // Move constructable
   NCCLComm(NCCLComm&& other) {
+    std::unique_lock<std::mutex> lock(other.mutex_);
```
Lets also add a comment here about why we have this locking behavior.

zhaojuanmao committed:
Update on "add lock for ncclCommAbort"


ncclCommAbort is not thread safe, so adding a lock for it

Differential Revision: [D19293869](https://our.internmc.facebook.com/intern/diff/D19293869/)

[ghstack-poisoned]
